### PR TITLE
p2p: bringing rlpx package updates

### PR DESCRIPTION
**Pull request by yoomee1313**
## Proposed changes
- This PR brings ethereum rlpx package updates. Listed on this issue: https://github.com/klaytn/klaytn/issues/1599
- In network perspective, there's application layer and transport layer. rlpx is included in transport layer.
  - application layer: dial(register statics), peer.addPeer/removePeer, read/write loops
  - transport layer: handshaking, read/write io interface, rlpx crypto encoding/decoding
- Those chages are about transport layer code refactoring, and main change is to clarify rlpx frame read/write code.
  - rlpx(rlpx.go) -> rlpxTransport(transport.go)
  - rlpxFrameRW struct (rlpx.go) -> conn struct (rlpx.go), sessionState(rlpx.go)
  - new snappy buffer library to increase the read/write speed of snappy buffer.
  - discReason type is changed from uint to uint8 to 

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [ ] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [ ] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: networks/p2p/server.go [hunk 1]
```diff
@@ -499,7 +498,18 @@ func (srv *MultiChannelServer) SetupConn(fd net.Conn, flags connFlag, dialDest *
...
 	}
 
-	c := &conn{fd: fd, transport: srv.newTransport(fd), flags: flags, conntype: common.ConnTypeUndefined, cont: make(chan error), portOrder: PortOrderUndefined}
+	// retrieve pubkey. if err occurs, dialPubkey is automatically set as nil
+	var dialPubkey *ecdsa.PublicKey
+	if dialDest != nil {
+		dialPubkey, _ = dialDest.ID.Pubkey()
+	}
+
+	c := &conn{fd: fd, flags: flags, conntype: common.ConnTypeUndefined, cont: make(chan error), portOrder: PortOrderUndefined}
```
Comment (sjnam) [inline networks/p2p/server.go:507]: If we move this line above line 501, we can simplify the if statement to check for `dialDest` is `nil`.

File: networks/p2p/server.go [hunk 2]
```diff
@@ -499,7 +498,18 @@ func (srv *MultiChannelServer) SetupConn(fd net.Conn, flags connFlag, dialDest *
 		return errors.New("shutdown")
 	}
 
-	c := &conn{fd: fd, transport: srv.newTransport(fd), flags: flags, conntype: common.ConnTypeUndefined, cont: make(chan error), portOrder: PortOrderUndefined}
+	// retrieve pubkey. if err occurs, dialPubkey is automatically set as nil
+	var dialPubkey *ecdsa.PublicKey
+	if dialDest != nil {
+		dialPubkey, _ = dialDest.ID.Pubkey()
```
Comment (sjnam) [inline networks/p2p/server.go:504]: Obtaining the `dialPubKey` is calculated twice in the `setupConn()` function.

File: networks/p2p/server.go (hunk 1 above)
Comment (yoomee1313) [reply networks/p2p/server.go:501]: Yes, refactored at https://github.com/klaytn/klaytn/pull/1954/commits/8e6351fe338cdd3575aacba14bf3120d121ee966.

File: networks/p2p/server.go (hunk 2 above)
Comment (yoomee1313) [reply networks/p2p/server.go]: Actually, dialPubkey derivation at `setupConn` is not needed anymore. So removed at https://github.com/klaytn/klaytn/pull/1954/commits/8e6351fe338cdd3575aacba14bf3120d121ee966.

**Review by sjnam — APPROVED** [review]

File: networks/p2p/peer_test.go [hunk 3]
```diff
@@ -87,7 +99,7 @@ func testPeerWithRWs(protos []Protocol, channelSize int) (func(), []*conn, *Peer
 	}
 
 	peer, _ := newPeer(serverSideConn, protos, defaultRWTimerConfig)
-	errc := make(chan error, 1)
+	errc := make(chan error, 2)
```
Comment (2dvorak) [inline networks/p2p/peer_test.go:102]: I cannot see why the size of this channel should be 2.. could you explain please?

Comment (yoomee1313) [reply networks/p2p/peer_test.go]: It was my mistake. There's no reason to change the size of the err channel. Applied at here https://github.com/klaytn/klaytn/pull/1954/commits/407e6c73512663a98a0edc7fe9a7bbac9d64797e.

**Review by 2dvorak — APPROVED** [review]

**Review by sjnam — APPROVED** [review]

**Pull request merged**